rpc-client-api/src/filter.rs

```diff
@@ -95,111 +77,115 @@ impl RpcFilterType {
     Base64DecodeError(#[from] base64::DecodeError),
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub enum MemcmpEncoding {
-    Binary,
-}
-
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
-#[serde(rename_all = "camelCase", untagged)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
+#[serde(rename_all = "camelCase", tag = "encoding", content = "bytes")]
 pub enum MemcmpEncodedBytes {
-    #[deprecated(
-        since = "1.8.1",
-        note = "Please use MemcmpEncodedBytes::Base58 instead"
-    )]
-    Binary(String),
     Base58(String),
     Base64(String),
     Bytes(Vec<u8>),
 }
 
+impl<'de> Deserialize<'de> for MemcmpEncodedBytes {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(untagged)]
+        enum DataType {
+            Encoded(String),
+            Raw(Vec<u8>),
+        }
+
+        #[derive(Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        enum RpcMemcmpEncoding {
+            Base58,
+            Base64,
+            Bytes,
+        }
+
+        #[derive(Deserialize)]
+        struct RpcMemcmpInner {
+            bytes: DataType,
+            encoding: Option<RpcMemcmpEncoding>,
+        }
+
+        let data = RpcMemcmpInner::deserialize(deserializer)?;
+
+        let memcmp_encoded_bytes = match data.bytes {
+            DataType::Encoded(bytes) => match data.encoding.unwrap_or(RpcMemcmpEncoding::Base58) {
+                RpcMemcmpEncoding::Base58 => MemcmpEncodedBytes::Base58(bytes),
+                RpcMemcmpEncoding::Base64 => MemcmpEncodedBytes::Base64(bytes),
+                _ => unreachable!(),
+            },
+            DataType::Raw(bytes) => MemcmpEncodedBytes::Bytes(bytes),
+        };
+
+        Ok(memcmp_encoded_bytes)
+    }
+}
+
 #[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
-#[serde(into = "RpcMemcmp", from = "RpcMemcmp")]
 pub struct Memcmp {
     /// Data offset to begin match
-    #[deprecated(
-        since = "1.15.0",
-        note = "Field will be made private in future. Please use a constructor method instead."
-    )]
-    pub offset: usize,
-    /// Bytes, encoded with specified encoding, or default Binary
-    #[deprecated(
-        since = "1.15.0",
-        note = "Field will be made private in future. Please use a constructor method instead."
-    )]
-    pub bytes: MemcmpEncodedBytes,
-    /// Optional encoding specification
-    #[deprecated(
-        since = "1.11.2",
-        note = "Field has no server-side effect. Specify encoding with `MemcmpEncodedBytes` variant instead. \
-            Field will be made private in future. Please use a constructor method instead."
-    )]
-    pub encoding: Option<MemcmpEncoding>,
+    offset: usize,
+    /// Bytes, encoded with specified encoding
+    #[serde(flatten)]
```

Comment:
TIL `serde(flatten)` is like:

```ts
{
    offset: number,
    ...bytes,
}
```

I see now how this isn't a breaking change from the perspective of the RPC API, because `bytes: {encoding: ... , bytes: ...}` gets spread into the `Memcmp` struct.

Reply:
Yeah, serde flatten enables lots of magicky stuff 😝